flytekit/extras/tensorflow/record.py

```diff
@@ -0,0 +1,146 @@
+import os
+from dataclasses import dataclass
+from typing import Generic, List, Optional, Tuple, Type, TypeVar, Union
+
+import tensorflow as tf
+from dataclasses_json import dataclass_json
+from tensorflow.core.example import example_pb2
+from tensorflow.python.data.ops.readers import TFRecordDatasetV2
+from typing_extensions import Annotated, get_args, get_origin
+
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+from flytekit.types.directory import TFRecordsDirectory
+from flytekit.types.file import TFRecordFile
+
+T = TypeVar("T")
+
+
+@dataclass_json
+@dataclass
+class TFRecordDatasetConfig:
+    """
+    TFRecordDatasetConfig can be used while creating tf.data.TFRecordDataset comprising
+    record of one or more TFRecord files.
+
+    Args:
+      compression_type: A scalar evaluating to one of "" (no compression), "ZLIB", or "GZIP".
+      buffer_size: The number of bytes in the read buffer. If None, a sensible default for both local and remote file systems is used.
+      num_parallel_reads: The number of files to read in parallel. If greater than one, the records of files read in parallel are outputted in an interleaved order.
+      name: A name for the operation.
+    """
+
+    compression_type: Optional[str] = None
+    buffer_size: Optional[int] = None
+    num_parallel_reads: Optional[int] = None
+    name: Optional[str] = None
+
+
+def extract_metadata(t: Type[T]) -> Tuple[T, TFRecordDatasetConfig]:
+    metadata = TFRecordDatasetConfig()
+    if get_origin(t) is Annotated:
+        base_type, metadata = get_args(t)
+        if isinstance(metadata, TFRecordDatasetConfig):
+            return base_type, metadata
+        else:
+            raise TypeTransformerFailedError(f"{t}'s metadata needs to be of type TFRecordDatasetConfig")
+    return t, metadata
+
+
+class TensorflowRecordTransformerBase(TypeTransformer, Generic[T]):
```

Comment:
Can we have two separate transformers for the file and directory? That would facilitate the maintenance of code.
cc: @cosmicBboy

Reply:
You can, however, include helper functions in case they need to be used in both transformers.